### PR TITLE
[ML] adding prediction_field_type to inference config

### PR DESCRIPTION
**Pull request by benwtrent**
Data frame analytics dynamically determines the classification field type. This field type then dictates the encoded JSON that is written to Elasticsearch. 

Inference needs to know about this field type so that it may provide the EXACT SAME predicted values as analytics. 

Here is added a new field `prediction_field_type` which indicates the desired type. Options are: `string` (DEFAULT), `number`, `boolean` (where close_to(1.0) == true, false otherwise). 

Analytics provides the default `prediction_field_type` when the model is created from the process.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java [hunk 1]
```diff
@@ -255,14 +259,36 @@ private InferenceConfig buildInferenceConfig(TargetType targetType) {
...
+                throw ExceptionsHelper.serverError(
                     "process created a model with an unsupported target type [{}]",
                     null,
-                    targetType));
-                return null;
+                    targetType);
         }
     }
 
+    PredictedFieldType getPredictedFieldType(Classification classification) {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsResultProcessor.java:269]: One suggestion that would help keeping the strings "string", "int", "bool" in one class would be as follows:
1. Make `Classification.getPredictionFieldType` method return the new `PredictedFieldType` enum.
2. Here (i.e. in `AnalyticsResultProcessor`) remove the translation in lines 277-285 (maybe keep the null -> STRING translation)
3. There (i.e. in `Classification`) in `getParams` method translate the enum to appropriate string ("string", "int", "bool").

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictedFieldType.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        return in.readEnum(PredictedFieldType.class);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeEnum(this);
+    }
+
+    @Override
+    public String toString() {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictedFieldType.java:37]: I prefer to keep public static methods near the top and member methods together.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictedFieldType.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    public Object transformPredictedValue(Double value, String stringRep) {
+        if (value == null) {
+            return null;
+        }
+        if(this == STRING) {
+            return stringRep == null ? value.toString() : stringRep;
+        }
+        if(this == BOOLEAN) {
+            return (1.0 - value) < EPS;
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictedFieldType.java]: added boundary checks and an exception if it is out of bounds.

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 4]
```diff
@@ -382,6 +382,10 @@ include::{docdir}/ml/ml-shared.asciidoc[tag=inference-config-results-field]
 `top_classes_results_field`::::
 (Optional, string)
 include::{docdir}/ml/ml-shared.asciidoc[tag=inference-config-classification-top-classes-results-field]
+
+`prediction_field_type`::::
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc:386]: > prediction_field_type

Not a big deal, but these properties are typically sorted alphabetically, and this seems to fall out of that sort order.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictionFieldType.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        }
+        switch(this) {
+            case STRING:
+                return stringRep == null ? value.toString() : stringRep;
+            case BOOLEAN:
+                if ((areClose(value, 1.0D) || areClose(value, 0.0D)) == false) {
+                    throw new IllegalArgumentException(
+                        "Cannot transform numbers other than 0.0 or 1.0 to boolean. Provided number [" + value + "]");
+                }
+                return (1.0D - value) <= EPS;
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictionFieldType.java:58]: Could `areClose` be used here as well for clarity?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictionFieldType.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Locale;
+
+/**
+ * The type of the predicted field.
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/PredictionFieldType.java:17]: s/predicted/prediction/
?

**Review by przemekwitek — APPROVED** [review]
LGTM

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**